### PR TITLE
feat(py): Expose dynamic rule condition validation

### PR DESCRIPTION
**Pull request by RaduW**
Added dynamic rule condition validation function in sentry-relay.

#skip-changelog

### CODE REVIEW

**Review by jan-auer — DISMISSED** [review]

File: py/sentry_relay/processing.py [hunk 1]
```diff
@@ -204,3 +205,17 @@ def parse_release(release):
     return json.loads(
         decode_str(rustcall(lib.relay_parse_release, encode_str(release)), free=True)
     )
+
+
+def validate_dynamic_rule_condition(condition):
```
Comment (jan-auer) [inline py/sentry_relay/processing.py]: Can we validate the entire `SamplingConfig` object instead of just the condition? Otherwise, there could be errors outside of the rules.

File: py/tests/test_processing.py [hunk 2]
```diff
@@ -173,3 +173,43 @@ def test_parse_release():
...
+    [
+        '{"op": "eq", "name": "field_1", "value": ["UPPER", "lower"], "options":{ "ignoreCase": true}}',
+        '{"op": "eq", "name": "field_2", "value": ["UPPER", "lower"]}',
+        '{"op": "glob", "name": "field_3", "value": ["1.2.*", "2.*"]}',
+        '{"op": "not", "inner": {"op": "glob", "name": "field_4", "value": ["1.*"]}}',
+        '{"op": "and", "inner": [{"op": "glob", "name": "field_5", "value": ["2.*"]}]}',
+        '{"op": "or", "inner": [{"op": "glob", "name": "field_6", "value": ["3.*"]}]}',
+        '{"op": "custom", "name": "event:legacy_browser", "value": ["ie9", "ie10", "ie11"]}',
+    ],
+    ids=("eq with case", "eq", "glob", "not", "and", "or", "custom"),
```
Comment (jan-auer) [inline py/tests/test_processing.py]: All of these should be unit tested. Let's only leave a single test here to test the binding.

File: relay-cabi/src/processing.rs [hunk 3]
```diff
@@ -220,3 +221,22 @@ pub unsafe extern "C" fn relay_parse_release(value: *const RelayStr) -> RelayStr
...
+/// volatile and need constant updating to keep in sync between Relay and Sentry
+#[no_mangle]
+#[relay_ffi::catch_unwind]
+pub unsafe extern "C" fn relay_validate_dynamic_rule_condition(value: *const RelayStr) -> RelayStr {
+    let ret_val = match serde_json::from_str::<RuleCondition>((*value).as_str()) {
+        Ok(condition) => {
+            if condition.supported() {
+                "".to_string()
+            } else {
+                "invalid condition".to_string()
```
Comment (jan-auer) [inline relay-cabi/src/processing.rs]: This should probably be `"unsupported condition"`?

File: relay-cabi/src/processing.rs [hunk 4]
```diff
@@ -220,3 +221,22 @@ pub unsafe extern "C" fn relay_parse_release(value: *const RelayStr) -> RelayStr
...
+/// Validate a dynamic rule condition.
+/// We only validate conditions in Relay since conditions are the only part of the rules that are
+/// volatile and need constant updating to keep in sync between Relay and Sentry
+#[no_mangle]
+#[relay_ffi::catch_unwind]
+pub unsafe extern "C" fn relay_validate_dynamic_rule_condition(value: *const RelayStr) -> RelayStr {
+    let ret_val = match serde_json::from_str::<RuleCondition>((*value).as_str()) {
+        Ok(condition) => {
+            if condition.supported() {
+                "".to_string()
```
Comment (jan-auer) [inline relay-cabi/src/processing.rs:232]: You can return `RelayStr::new("")` here. No allocation required.

File: py/sentry_relay/processing.py (hunk 1 above)
Comment (RaduW) [reply py/sentry_relay/processing.py]: We certainly can. 
I thought about this for a bit. 
In the end I decided that the best thing to do is validate the most "volatile" part in Rust, and the rest in Python.

The reason for this is that using the DjangoRest validators gives a very precise location and structure to the validation error. I could replicate this structure in Rust but it would seem strange and disconnected (and some additional work).

So to get the best of both worlds: precise location of error and the confidence that the validator is kept in sync, I opted for this "mixed" approach.

File: py/tests/test_processing.py (hunk 2 above)
Comment (RaduW) [reply py/tests/test_processing.py]: The operators are unit tested. 

What I test here is that the validator works as expected, not the functionality of the operator and not even the serialization of the structure (although the validator effectively uses serialization to do its work).

File: relay-cabi/src/processing.rs (hunk 3 above)
Comment (RaduW) [reply relay-cabi/src/processing.rs]: yeah, that's a better error, changed.

File: relay-cabi/src/processing.rs (hunk 4 above)
Comment (RaduW) [reply relay-cabi/src/processing.rs:232]: I know, but I don't like it.
If you **insist** I'll do it but here are my reasons not to:

What I want to do (in a future PR) is change the current `RealyStr::new(s: &str)` to:
```
   pub(crate) fn new(s: &'static str) -> RelayStr 
```
and
```
   pub(crate) fn unsafe unsafe_new(s: &str) -> RelayStr 
```

That also affects the implementation of ` impl<'a> From<&'a str> for RelayStr` so I need to look
a bit more in detail at it.

Until then I consider the implementation of `RealyStr::new` as broken, it is unsafe and breaks the Rust
guarantees. It will happily compile something like:

```
pub fn bad_function() -> RelayStr {
    RelayStr::new(&format!("Hello {}", "Jan"))
}
```

File: py/sentry_relay/processing.py (hunk 1 above)
Comment (jan-auer) [reply py/sentry_relay/processing.py]: Makes sense, using Django validators for precise location is good. You could still have librelay validate the entire configuration structure to ensure we never pass invalid data on (especially when we make updates). The Python code in Sentry is too likely to go out of sync. Having both side-by-side should also work.

File: py/tests/test_processing.py (hunk 2 above)
Comment (jan-auer) [reply py/tests/test_processing.py]: To clarify: Let's please put these tests in Rust code, ideally into the module that declares the types inside `relay-sampling`. Then, only keep a single Python test here that asserts the wrapper works. There's no point in effectively testing `serde` derives through the python layer.

Comment (RaduW) [reply py/tests/test_processing.py]: got it

File: relay-cabi/src/processing.rs (hunk 4 above)
Comment (mitsuhiko) [reply relay-cabi/src/processing.rs:232]: @jan-auer i feel like we should revisit this string pattern. I remember we incorrectly deduced what the allocation / borrow behavior once before.

Comment (jan-auer) [reply relay-cabi/src/processing.rs:232]: @mitsuhiko Right, we had bugs prior to Relay in libsymbolic. It's worth noting that the Python layer **always** copies the string anyway. 

@RaduW you're right, the name `new` doesn't carry the implications the function has. It is worth noting that `RelayStr` is only to be used in an unsafe FFI layer, so the exposure is pretty low. Let's use `RelayStr::default` for now, which will inherently not have any borrowing problems.

To continue a discussion from Slack: There is not much code, and it is easy to enumerate _all_ places where we use `RelayStr::new`:
- `relay_data_category_name`, which always returns static strings
- `RelayStr::default`, which can be considered an implementation detail of empty strings in `RelayStr`
- `relay_str_from_cstr`, which deals with pointers. It's also dead code
- `relay_validate_pii_config`, which can also be changed to `RelayStr::default`
- `relay_convert_datascrubbing_config`, which returns a static `"{}"`.

To sum this up, we can address this much simpler by:
- Changing the signature to `new(s: &'static str) -> Self` (this is now inherently safe)
- Removing `relay_str_from_cstr`

File: py/sentry_relay/processing.py (hunk 1 above)
Comment (RaduW) [reply py/sentry_relay/processing.py]: Ok added a global validator (to validate the whole structure with a Yes/No , a condition validator to pinpoint problem conditions and will be using both in DjangoRest validators to get nice error messages

File: relay-cabi/src/processing.rs [hunk 5]
```diff
@@ -220,3 +221,46 @@ pub unsafe extern "C" fn relay_parse_release(value: *const RelayStr) -> RelayStr
...
+/// go out of sync.
+#[no_mangle]
+#[relay_ffi::catch_unwind]
+pub unsafe extern "C" fn relay_validate_sampling_configuration(value: *const RelayStr) -> RelayStr {
+    let ret_val = match serde_json::from_str::<SamplingConfig>((*value).as_str()) {
+        Ok(config) => {
+            let mut error_str: Option<String> = None;
+            for rule in config.rules {
+                if !rule.condition.supported() {
+                    error_str = Some("unsupported sampling rule".to_string());
```
Comment (jan-auer) [inline relay-cabi/src/processing.rs]: Can you just make this an early return? For example:

```rust
    match serde_json::from_str::<SamplingConfig>((*value).as_str()) {
        Ok(config) => {
            for rule in config.rules {
                if !rule.condition.supported() {
                    return RelayStr::new("unsupported sampling rule");
                }
            }
            
            RelayStr::default()
        }
        Err(e) => RelayStr::from_string(e.to_string()),
    }
```

Note that I'm using `new` here as per discussion in https://github.com/getsentry/relay/pull/941#discussion_r585374002.

**Review by jan-auer — APPROVED** [review]
Thanks, this is good to merge. 

A side note on naming: One mentions "dynamic rule", the other "sampling". Given that both validate within the same structure, we could choose more similar names, such as `validate_sampling_config` and `validate_sampling_condition`.
